Add Files, Clipboard, and first Advanced Apps Patterns (#7645)
GoogleChrome/web.dev: src/site/content/en/patterns/advanced-apps/audio-effects/index.md

---
layout: pattern
title: How to add effects to audio
date: 2022-09-01
authors:
  - conwayt
description: >
  Learn how to use the Web Audio API to add effects to audio sources.
height: 800
---

## The modern way
The Web Audio API uses the `AudioContext()` interface to manage sources, filters, and destinations. Once you’ve created a new `AudioContext()`, create an audio source node, like an `AudioBufferSourceNode` or `OscillatorNode`.  As an example, consider a basic oscillator with a [low pass filter](https://en.wikipedia.org/wiki/Low-pass_filter) applied.

{% BrowserCompat 'api.AudioContext' %}

> ### Using the `createBiquadFilter()` method

Reviewer: ```suggestion

### Using the `createBiquadFilter()` method
```

First, create a new `AudioContext()`. Then create an audio source node, like an [`AudioBufferSourceNode`](https://developer.mozilla.org/docs/Web/API/AudioBufferSourceNode) or [`OscillatorNode`](https://developer.mozilla.org/docs/Web/API/OscillatorNode).  You'll create a `sine` oscillator node, for this example, which has a frequency of 420 hertz from the moment it starts playing.

```js
const audioCtx = new (window.AudioContext || window.webkitAudioContext)();
const oscillator = audioCtx.createOscillator()
oscillator.type = 'sine';
 // Value is in hertz.
oscillator.frequency.setValueAtTime(420, audioCtx.currentTime);
```

> Next, create an effect node using `createBiquadFilter()`. Set the type to `lowpass` and the frequency to start 1 second after the start of playback. Then connect the `biquadFilter` to the `oscillator`.

Reviewer: ```suggestion
Next, create an effect node using `createBiquadFilter()`. Set the type to `lowpass` and the frequency to start one second after the start of playback. Then connect the `biquadFilter` to the `oscillator`.
```

```js
const biquadFilter = audioCtx.createBiquadFilter();
biquadFilter.type = 'lowpass';
biquadFilter.frequency.setValueAtTime(200, audioCtx.currentTime + 1);
oscillator.connect(biquadFilter);
```

> Finally connect the `biquadFilter` to the destination of the `audioCtx`, before starting the `oscillator` and stopping it after 2 seconds of playback.

Reviewer: ```suggestion
Finally connect the `biquadFilter` to the destination of the `audioCtx`, before starting the `oscillator` and stopping it after two seconds of playback.
```

```js
biquadFilter.connect(audioCtx.destination);
oscillator.start();
oscillator.stop(2);
```

> The sound from the oscillator is passed through the unaffected filter, and on to the destination, which is your computer's speakers. After 1 second of playback, the `lowpass` filter comes into effect. After 2 seconds, the oscillator is stopped.

Reviewer: ```suggestion
The sound from the oscillator is passed through the unaffected filter, and on to the destination, which is your computer's speakers. After one second of playback, the `lowpass` filter comes into effect. After two seconds, the oscillator is stopped.
```
